File: packages/ingest/src/lib/mgl/cross-references.ts
export interface SectionCrossReference {
	section: string;
	chapter: string;
	offset: number;
	length: number;
	link: string;
}

const CHAPTER_SECTION_RE =
	/\bchapter\s+(\d+[a-zA-Z]?)\s*,\s*section\s+(\d+[a-zA-Z]?)\b/gi;
const OF_CHAPTER_RE =
	/\bsection\s+(\d+[a-zA-Z]?)\s+of\s+chapter\s+(\d+[a-zA-Z]?)\b/gi;

function normalizeDesignator(value: string): string {
	return value.trim().toUpperCase();
}

function makeLink(chapter: string, section: string): string {
	return `/statutes/mgl/chapter/${chapter.toLowerCase()}/section/${section.toLowerCase()}`;
}

function collectMatches(
	text: string,
	regex: RegExp,
	map: (match: RegExpMatchArray) => { chapter: string; section: string } | null,
): SectionCrossReference[] {
	const refs: SectionCrossReference[] = [];
	for (const match of text.matchAll(regex)) {
		const full = match[0];
		if (!full || match.index === undefined) {
			continue;
		}
		const mapped = map(match);
		if (!mapped) continue;
		const chapterRaw = mapped.chapter;
		const sectionRaw = mapped.section;
		const chapter = normalizeDesignator(chapterRaw);
		const section = normalizeDesignator(sectionRaw);
		refs.push({
			chapter,
			section,
			offset: match.index,
			length: full.length,
			link: makeLink(chapter, section),
		});
	}
	return refs;
}

export function extractSectionCrossReferences(
	text: string,
): SectionCrossReference[] {
	const refs = [
		...collectMatches(text, CHAPTER_SECTION_RE, (match) => {
			const chapter = match[1];
			const section = match[2];
			if (!chapter || !section) return null;
			return { chapter, section };
		}),
		...collectMatches(text, OF_CHAPTER_RE, (match) => {
			const section = match[1];
			const chapter = match[2];
			if (!chapter || !section) return null;
			return { chapter, section };
		}),
	];

	const unique = new Map<string, SectionCrossReference>();
	for (const ref of refs) {
		const key = `${ref.chapter}:${ref.section}:${ref.offset}`;
		if (!unique.has(key)) unique.set(key, ref);
	}

	return [...unique.values()].sort((a, b) => a.offset - b.offset);
}

File: packages/ingest/container-rust/src/sources/mgl/cross_references.rs
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::sync::LazyLock;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SectionCrossReference {
    pub section: String,
    pub chapter: String,
    pub offset: usize,
    pub length: usize,
    pub link: String,
}

static CHAPTER_SECTION_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\bchapter\s+(\d+[a-zA-Z]?)\s*,\s*section\s+(\d+[a-zA-Z]?)\b")
        .expect("CHAPTER_SECTION_RE should compile")
});

static OF_CHAPTER_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\bsection\s+(\d+[a-zA-Z]?)\s+of\s+chapter\s+(\d+[a-zA-Z]?)\b")
        .expect("OF_CHAPTER_RE should compile")
});

fn normalize_designator(value: &str) -> String {
    value.trim().to_uppercase()
}

fn make_link(chapter: &str, section: &str) -> String {
    format!(
        "/statutes/chapter/{}/section/{}",
        chapter.to_lowercase(),
        section.to_lowercase()
    )
}

pub fn extract_section_cross_references(text: &str) -> Vec<SectionCrossReference> {
    let mut refs = Vec::new();

    // Pattern: "chapter X, section Y"
    for caps in CHAPTER_SECTION_RE.captures_iter(text) {
        let full = match caps.get(0) {
            Some(f) => f,
            None => continue,
        };
        let chapter = caps.get(1).map(|m| m.as_str()).unwrap_or("");
        let section = caps.get(2).map(|m| m.as_str()).unwrap_or("");
        if chapter.is_empty() || section.is_empty() {
            continue;
        }
        let chapter = normalize_designator(chapter);
        let section = normalize_designator(section);
        refs.push(SectionCrossReference {
            chapter: chapter.clone(),
            section: section.clone(),
            offset: full.start(),
            length: full.end() - full.start(),
            link: make_link(&chapter, &section),
        });
    }

    // Pattern: "section X of chapter Y"
    for caps in OF_CHAPTER_RE.captures_iter(text) {
        let full = match caps.get(0) {
            Some(f) => f,
            None => continue,
        };
        let section = caps.get(1).map(|m| m.as_str()).unwrap_or("");
        let chapter = caps.get(2).map(|m| m.as_str()).unwrap_or("");
        if chapter.is_empty() || section.is_empty() {
            continue;
        }
        let chapter = normalize_designator(chapter);
        let section = normalize_designator(section);
        refs.push(SectionCrossReference {
            chapter: chapter.clone(),
            section: section.clone(),
            offset: full.start(),
            length: full.end() - full.start(),
            link: make_link(&chapter, &section),
        });
    }

    // Dedupe
    let mut seen = std::collections::HashSet::new();
    let mut deduped = Vec::new();
    for reference in refs {
        let key = format!(
            "{}:{}:{}:{}",
            reference.chapter, reference.section, reference.offset, reference.link
        );
        if seen.insert(key) {
            deduped.push(reference);
        }
    }

    deduped.sort_by_key(|r| r.offset);
    deduped
}

/// Inlines cross-references as markdown links in the text
pub fn inline_section_cross_references(text: &str) -> String {
    let mut references = extract_section_cross_references(text);
    references.sort_by(|a, b| b.offset.cmp(&a.offset));

    let mut output = text.to_string();
    for reference in references {
        let start = reference.offset;
        let end = start.saturating_add(reference.length);
        if end > output.len()
            || start >= end
            || !output.is_char_boundary(start)
            || !output.is_char_boundary(end)
        {
            continue;
        }

        let label = &output[start..end];
        let linked = format!("[{label}]({})", reference.link);
        output.replace_range(start..end, &linked);
    }

    output
}
